server/monitor-types/tls.js

```diff
@@ -0,0 +1,266 @@
+                const dataString = data.toString().trim();
+                this.debug_log(`read data: '${dataString}'`);
+                cleanup();
+                resolve(dataString);
+            };
+
+            aborter.addEventListener("abort", onAbort, { once: true });
+
+            socket.on("error", onError);
+            socket.on("data", onData);
+            socket.resume();
+        });
+    }
+
+    /**
+     * Reads available data from the given socket if it starts with a given prefix.
+     * @param {AbortController} aborter  Abort controller used to abort the read
+     * @param {*}               socket   net.Socket or tls.TLSSocket instance to use
+     * @param {string}          expected Prefix the response is expected to start with
+     * @returns {Promise<string>} Data read from the socket or rejected promise if the response does
+     *                            not start with the prefix
+     */
+    async expectDataStartsWith(aborter, socket, expected) {
+        const data = await this.readData(aborter, socket);
+        this.debug_log(`response data: '${data}', expected: '${expected}'…`);
+        if (!data.startsWith(expected)) {
+            throw new Error(`Unexpected response. Data: '${data}', Expected: '${expected}'…`);
+        }
+    }
+
+    /**
+     * Performs STARTTLS on the given socket.
+     * @param {AbortController} aborter Abort controller used to abort the STARTTLS process
+     * @param {*}               socket  net.Socket or tls.TLSSocket instance to use
+     * @returns {Promise<void>} Rejected promise if the STARTTLS process failed
+     */
+    async startTls(aborter, socket) {
+        this.debug_log("waiting for prompt");
+        await this.expectDataStartsWith(aborter, socket, "220 ");
+        this.debug_log("sending STARTTLS");
+        socket.write("STARTTLS\n");
+        this.debug_log("waiting for ready-to-TLS");
+        await this.expectDataStartsWith(aborter, socket, "220 ");
+    }
+
+    /**
+     * Upgrades an unencrypted socket to a TLS socket using STARTTLS.
+     * @param {*}               socket  net.Socket representing the unencrypted connection
+     * @returns {Promise<TLSSocket>} tls.TLSSocket instance representing the upgraded TLS connection
+     *                               or rejected promise if the STARTTLS process failed
+     */
+    upgradeConnection(socket) {
+        return new Promise((resolve, reject) => {
+            const onSecureConnect = () => {
+                const cipher = tlsSocket.getCipher();
+                this.debug_log(`connected: authorized: ${tlsSocket.authorized}, cipher: ${cipher ? cipher.name : "<none>"}`);
+                resolve(tlsSocket);
+            };
+
+            const onError = (error) => {
+                reject(error);
+            };
+
+            const tlsSocket = tls.connect({ socket: socket });
+
+            tlsSocket.on("secureConnect", onSecureConnect);
+            tlsSocket.on("error", onError);
```

Comment:
is there a reason for this extra function?

```suggestion
            tlsSocket.on("error", reject);
```

Reply:
Good catch! I originally had some debug logging in the `onError` function. With that gone, the function is indeed unnecessary.